File: src/0107.binary-tree-level-order-traversal-ii.107/solution.go
package binarytreelevelordertraversalii107

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func levelOrderBottom(root *TreeNode) [][]int {
	result := [][]int{}
	if root == nil {
		return result
	}

	var bfs func([]*TreeNode)
	bfs = func(nodes []*TreeNode) {
		vals := []int{}
		next := []*TreeNode{}
		for _, node := range nodes {
			vals = append(vals, node.Val)
			if node.Left != nil {
				next = append(next, node.Left)
			}
			if node.Right != nil {
				next = append(next, node.Right)
			}
		}
		if len(next) > 0 {
			bfs(next)
		}
		result = append(result, vals)
	}
	bfs([]*TreeNode{root})
	return result
}

File: src/0107.binary-tree-level-order-traversal-ii.107/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        if (root == null) {
            return List.of();
        }
        List<List<Integer>> result = new LinkedList<>();

        travel(List.of(root), result);

        return result;
    }

    private void travel(List<TreeNode> level, List<List<Integer>> result) {
        List<Integer> row = new ArrayList<>();
        List<TreeNode> nextLevel = new ArrayList<>();
        for (TreeNode node : level) {
            row.add(node.val);
            if (node.left != null) {
                nextLevel.add(node.left);
            }
            if (node.right != null) {
                nextLevel.add(node.right);
            }
        }
        result.addFirst(row);
        if (!nextLevel.isEmpty()) {
            travel(nextLevel, result);
        }

    }
}